### PR TITLE
BUG: sparse.csgraph.dijkstra: fix dtype and shape bugs

### LINKED ISSUE #20904: BUG: sparse.csgraph.dijkstra errors on inputs with int64 or no indices
Two little bugs I found on the latest pip-pre release (probably in the current release, too, I imagine) on macOS arm64 Python 3.12 at least are below. 

Starting with with imports and sparse matrix creation:
```
import numpy as np
from scipy import sparse
import traceback

adj_mat = sparse.eye(4, format="csr")
```

# Works okay
```
sparse.csgraph.dijkstra(
    adj_mat, False, indices=[1], min_only=True, return_predecessors=True
)
```

# BUG 1: Changing to default indices=None emits a cryptic error

```
sparse.csgraph.dijkstra(
    adj_mat, False, min_only=True, return_predecessors=True
)
#   File "_shortest_path.pyx", line 634, in scipy.sparse.csgraph._shortest_path.dijkstra
# ValueError: cannot reshape array of size 4 into shape (4,4)
```

# BUG 2: doesn't work with int64 indices

```
adj_mat = sparse.csr_array((adj_mat.data, adj_mat.indices.astype(np.int64), adj_mat.indptr.astype(np.int64)))
sparse.csgraph.dijkstra(
    sparse.csr_matrix(adj_mat), False, indices=[1], min_only=True, return_predecessors=True
)
#   File "_shortest_path.pyx", line 620, in scipy.sparse.csgraph._shortest_path.dijkstra
# ValueError: Buffer dtype mismatch, expected 'const int' but got 'long'
```

### PR DESCRIPTION
**Pull request by larsoner**
Closes #20904

In theory maybe fused dtypes or buffer protocol or something could be used here. And `bellman_ford` and `johnson` might have the same issues, I haven't checked.

### CODE REVIEW

File: scipy/sparse/csgraph/tests/test_shortest_path.py [hunk 1]
```diff
@@ -33,9 +33,9 @@
 directed_2SP_0_to_3 = [[-9999, 0, -9999, 1, -9999],
                        [-9999, 0, -9999, 4, 1]]
 
-directed_sparse_zero_G = scipy.sparse.csr_matrix(([0, 1, 2, 3, 1], 
-                                            ([0, 1, 2, 3, 4], 
-                                             [1, 2, 0, 4, 3])), 
+directed_sparse_zero_G = scipy.sparse.csr_matrix(([0, 1, 2, 3, 1],
+                                            ([0, 1, 2, 3, 4],
+                                             [1, 2, 0, 4, 3])),
```
Comment (lucascolley) [inline scipy/sparse/csgraph/tests/test_shortest_path.py:38]: optionally, we could fix the indentation here, while the lines are already being touched.

File: scipy/sparse/csgraph/tests/test_shortest_path.py [hunk 2]
```diff
@@ -44,9 +44,9 @@
                       [np.inf, np.inf, np.inf, 0, 3],
                       [np.inf, np.inf, np.inf, 1, 0]]
 
-undirected_sparse_zero_G = scipy.sparse.csr_matrix(([0, 0, 1, 1, 2, 2, 1, 1], 
-                                              ([0, 1, 1, 2, 2, 0, 3, 4], 
-                                               [1, 0, 2, 1, 0, 2, 4, 3])), 
+undirected_sparse_zero_G = scipy.sparse.csr_matrix(([0, 0, 1, 1, 2, 2, 1, 1],
+                                              ([0, 1, 1, 2, 2, 0, 3, 4],
+                                               [1, 0, 2, 1, 0, 2, 4, 3])),
```
Comment (lucascolley) [inline scipy/sparse/csgraph/tests/test_shortest_path.py:49]: likewise here

**Review by lucascolley — APPROVED** [review]
this LGTM, thanks @larsoner ! I checked that the new test fails on `main`.

**Comment by larsoner** [comment]
@lucascolley pushed an indentation change

**Review by lucascolley — APPROVED** [review]
thanks again!

**Pull request merged**